### DISCUSSION TITLE
Replace reference object type with another schema path

### DISCUSSION THREAD

**Discussion by PintOfBitter** [victools/jsonschema-generator#436] [Q&A] [answered]
Hello,

I am having trouble generating a specific schema. Let's say I have the following classes:

**package org.example.geo:**
```java
public class Region {
    private int region_id;
    private String region_name;
}
```
and
**package org.example.studentaffairs:**
```java
public class Student {
    private int student_id;
    private String student_name;
    private Region region;
}
```
What I am trying to achieve are the following schemas:
for region:
```json
{
    "type": "object",
    "properties": {
        "region_id": {
            "type" : "int"
        },
        "region_name": {
            "type": "String"
        }
    }
}
```
This will be saved under org.example.geo.region
And for student:
```json
    { 
        "type": "object",
        "properties": {
            "student_id": {
                "type" : "int"
            },
            "student_name": {
                "type": "String"
            },
            "region": {
                "type": "org.example.geo.region"
            }
        }
    }
```

The part I am failing to do is generating region's type as "org.example.geo.region".
Is there a way I can access the objects being generated and replace them with a reference with a path to the referenced object?

Any help or clue is greatly appreciated!

Thanks,
PoB

**Comment by CarstenWickner** [comment, marked as answer]
Hi @PintOfBitter,

what you're proposing is strictly speaking an invalid JSON schema. There are only a handful of supported values for the `type` keyword. https://json-schema.org/understanding-json-schema/reference/type
It's not meant to be overridden with custom references to external schemas. The `$ref` keyword supports that kind of thing.

----

Since others have asked for similar things, there is already an example for introducing external references here: [ExternalRefPackageExample](https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/main/java/com/github/victools/jsonschema/examples/ExternalRefPackageExample.java)
If you search for it here on the repository, you'll see the few issues/discussions raised on that topic.

Your target schema file for the `student` would then more look like this:
```json
{ 
    "type": "object",
    "properties": {
        "student_id": {
            "type" : "integer"
        },
        "student_name": {
            "type": "string"
        },
        "region": {
            "$ref": "./org/example/geo/region"
        }
    }
}
```

---

However, for completeness' sake: technically you can override anything – including the `type` keyword – with what I'm calling a [TypeAttributeOverride](https://victools.github.io/jsonschema-generator/#type-attribute-overrides).
Adapting the example from the documentation, you could do something like this (obviously with some more sophisticated logic to make it reusable):
```java
configBuilder.forTypesInGeneral()
    .withTypeAttributeOverride((node, scope, context) -> {
            if (scope.getType().getErasedType() == Region.class) {
                node.put(context.getKeyword(SchemaKeyword.TAG_TYPE), "org.example.geo.region");
            }
    });
```

**Reply by PintOfBitter** [reply]
Hello @CarstenWickner,

Greatly appreciate the quick response! And thanks for pointing out the issue with changing the "type" value, this is especially helpful for someone still learning the ropes on Json schemas.

I will go over the example shared and reference external schemas through the $ref.
